### PR TITLE
Typescript saga immutablejs

### PR DESCRIPTION
**Pull request by bflamand**
### PR checklist
 
- [ ] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [ ] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) beforehand.
- [ ] Run the shell script `./bin/generate-samples.sh`to update all Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. These must match the expectations made by your contribution. You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ ] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`
- [ ] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java [hunk 1]
```diff
@@ -38,6 +38,11 @@
      * may be represented as 'int', 'int32', 'Integer', etc, depending on the programming language.
      */
     public String dataType;
+    /**
+     * Alternative version of the language-specific data type when generator needs a different output in some templates.
+     * For example this is used by typescript fetch generator to output either Array (javascript) or List (immutablejs)
+     */
+    public String dataTypeAlternate;
```
Comment (JFCote) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java]: You should not do this. This is a common thing to do in most of the existing template. You need to deal with this in the TypescriptFetchCodegen.java. You can you typeMapping or even postProcessing. But don't create yet another property

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java [hunk 2]
```diff
@@ -108,6 +113,15 @@
...
     public boolean isPrimitiveType;
     public boolean isModel;
+    /**
+     * Is a model containing an "id" property marked as isUniqueId; field populated only by typescript fetch for now.
+     */
+    public boolean isEntity;
+    /**
+     * The property represents a unique id (x-isUniqueId: true); field populated only by typescript fetch for now.
+     */
+    public boolean isUniqueId;
```
Comment (JFCote) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java]: is there a common way to name  properties that are on the vendor side so people are not tempted to use them in their template by error?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java [hunk 3]
```diff
@@ -158,6 +172,11 @@
     // If 'additionalProperties' is set to a type or refers to a type, 'items' provides the type information for
     // the undeclared properties. 
     public CodegenProperty items;
+
+    public String itemsDataType; // if isContainer == true, this is the dataType of the items else it is null. Used by typescript fetch only for now.
```
Comment (JFCote) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java]: Don't use these 3 tags since they have duplicated information.
itemsDataType: This is already in the item, you don't need this.
itemsAreModels: You can do "if" in your template based on the isUniqueId
itemsAreEntities: You can do "if" in your template based on the item type.  This is very common in most template.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java [hunk 4]
```diff
@@ -5450,12 +5450,7 @@ public void updateCodegenPropertyEnum(CodegenProperty var) {
         // handle default value for enum, e.g. available => StatusEnum.AVAILABLE
         if (var.defaultValue != null) {
             String enumName = null;
-            final String enumDefaultValue;
-            if ("string".equalsIgnoreCase(dataType)) {
-                enumDefaultValue = toEnumValue(var.defaultValue, dataType);
-            } else {
-                enumDefaultValue = var.defaultValue;
-            }
+			final String enumDefaultValue = var.defaultValue;
```
Comment (JFCote) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: Why did you remove this?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java [hunk 5]
```diff
@@ -621,11 +621,6 @@ public String toEnumValue(String value, String datatype) {
         }
     }
 
-    @Override
```
Comment (JFCote) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java]: same question here, why did you remove this?

File: modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    {{#lambda.camelcase}}{{.}}{{/lambda.camelcase}}RecordUtils,
+{{/modelImports}}
+} from './';
+{{/hasImports}}
+
+export const {{classname}}RecordProps = {
+	recType: "{{classname}}Record" as "{{classname}}Record",
+{{#vars}}
+    {{#isArray}}
+    {{#itemsAreModels}}
```
Comment (JFCote) [inline modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache]: This is directly related to what I have said above, you don't need these 3 new properties.
The logic is: 
* All items are models unless they are native type
* You can know that a model is an "entity" if isUnique is true.

File: modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+export type {{classname}}RecordEntity = RecordOf<{{classname}}RecordEntityPropsType>;
+
+knownRecordFactories.set({{classname}}RecordEntityProps.recType, {{classname}}RecordEntity);
+{{/isEntity}}
+
+class {{classname}}RecordUtils extends ApiRecordUtils<{{classname}}, {{classname}}Record> {
+	public normalize(apiObject: {{classname}}, asEntity?: boolean): {{classname}} {
+		(apiObject as any).recType = {{#isEntity}}asEntity ? "{{classname}}RecordEntity" : {{/isEntity}}"{{classname}}Record";
+{{#vars}}
+        {{#isUniqueId}}
```
Comment (JFCote) [inline modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache:98]: you are already doing it the right way here

File: modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            {{#lambda.camelcase}}{{name}}{{/lambda.camelcase}}: [{{#lambda.camelcase}}{{itemsDataType}}{{/lambda.camelcase}}Utils.getSchema()],
+            {{/itemsAreEntities}}
+{{/vars}}
+		});
+	}
+{{/isEntity}}
+{{#isMetaDataResponse}}
+{{#vars.1}}
+
+	public fromApi_data(apiObject: {{classname}}): {{{dataTypeAlternate}}} {
```
Comment (JFCote) [inline modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache]: like I said above, all this thing with the dataTypeAlternate must be  handled in the codegen code

File: modules/openapi-generator/src/main/resources/typescript-fetch/runtimeSagasAndRecords.mustache [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+import {fromJS as originalFromJS, isIndexed, List, Map as ImmMap, RecordOf} from 'immutable';
+import {normalize, NormalizedSchema, schema, Schema} from "normalizr";
+import {ActionDefinition, createAction} from "redux-ts-simple";
+
+export const knownRecordFactories = new Map<string, any>();
+export const knownIndexedSetByKey = [];
+
+export function appFromJS(any: any): any {
+	return originalFromJS(any, (key, value) => {
```
Comment (JFCote) [inline modules/openapi-generator/src/main/resources/typescript-fetch/runtimeSagasAndRecords.mustache]: please review all your file and make sure you use a "4 space tabs" instead of "real tabs". this is the standard in this project

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenOperation.java [hunk 10]
```diff
@@ -29,9 +29,9 @@
             isArray, isMultipart,
             isResponseBinary = false, isResponseFile = false, hasReference = false,
             isRestfulIndex, isRestfulShow, isRestfulCreate, isRestfulUpdate, isRestfulDestroy,
-            isRestful, isDeprecated, isCallbackRequest, uniqueItems;
-    public String path, operationId, returnType, returnFormat, httpMethod, returnBaseType,
-            returnContainer, summary, unescapedNotes, notes, baseName, defaultResponse; 
+            isRestful, isDeprecated, isCallbackRequest, uniqueItems, returnTypeIsMetaDataResponse, returnTypeIsMetaOnlyResponse, returnTypeSupportsEntities, returnTypeIsModel, returnTypeIsArray;
```
Comment (JFCote) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenOperation.java]: it kind of weird to have a concept of entity in a very generic object called CodegenOperation. I think this should be handled in the specific codegen in the java code. I don't think any of these variable should be here.

Comment (JFCote) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenOperation.java]: Maybe follow the pattern here?
`class ExtendedCodegenOperation extends CodegenOperation {`

**Review by JFCote — CHANGES_REQUESTED** [review]
Lot of things to remove and/or move

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java (hunk 1 above)
Comment (bflamand) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java]: From what I understand, the only way to have another property accessible in the mustache file for models is to add it here? If there was another way, this would be great, we would not have to "polute" global codegen here...
Btw, this would not be needed if we made a new generator specifically for outputing sagas&records. But here we need to continue to output the normal typescript fetch in addition to the new stuff so we need a second and different dataType here for the new stuff. This would also not be needed if mustache fully supported if/else but it does not. So we need to do the if/else in code and prepare this variable so it can be used in records and saga mustache files.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java (hunk 2 above)
Comment (bflamand) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java]: I looked and did not find a pattern for this. Naming seams to be different depending on who added a variable. But we could at least use one for our code and remain consistant.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java (hunk 4 above)
Comment (bflamand) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java]: I found a bug!

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java (hunk 3 above)
Comment (bflamand) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java]: again, this was added because I did not find any way to get this information once in the mustache file. I needed to expose it here.

File: modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache (hunk 6 above)
Comment (bflamand) [reply modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache]: Will check if {{^nativeType}} is the same as itemsAreModels ...

File: modules/openapi-generator/src/main/resources/typescript-fetch/runtimeSagasAndRecords.mustache (hunk 9 above)
Comment (bflamand) [reply modules/openapi-generator/src/main/resources/typescript-fetch/runtimeSagasAndRecords.mustache]: oh...
